python/ray/experimental/data/dataset.py

```diff
@@ -799,14 +799,81 @@ def iter_batches(self,
             prefetch_blocks: The number of blocks to prefetch ahead of the
                 current block during the scan.
             batch_size: Record batch size, or None to let the system pick.
-            batch_format: Specify "pandas" to select ``pandas.DataFrame`` as
-                the batch format, or "pyarrow" to select ``pyarrow.Table``.
+            batch_format: The format in which to return each batch.
+                Specify "pandas" to select ``pandas.DataFrame``, "pyarrow" to
+                select ``pyarrow.Table``, or "blocks" to return the raw block.
+                Default is "pandas".
+            drop_last: Whether to drop the last batch if it's incomplete.
 
         Returns:
             A list of iterators over record batches.
         """
 
-        raise NotImplementedError  # P1
+        def chunk(iterable, n):
+            it = iter(iterable)
+            while True:
+                chunk_it = itertools.islice(it, n)
+                try:
+                    first = next(chunk_it)
+                except StopIteration:
+                    return
+                yield itertools.chain((first, ), chunk_it)
+
+        def format_batch(batch, format):
+            if batch_format == "pandas":
+                return batch.to_pandas()
+            elif batch_format == "pyarrow":
+                return batch._table
+            elif batch_format == "blocks":
+                return batch
+            else:
+                raise ValueError(
+                    f"The given batch format: {batch_format} "
+                    f"is invalid. Supported batch type: {BatchType}")
+
+        batch_buffer = None
+        for block_chunk in chunk(self._blocks, prefetch_blocks + 1):
+            block_chunk = list(block_chunk)
```

Comment:
Is this a chunk or sliding window? The previous code was easier to read

Reply:
Chunks would also not be prefetching across chunk boundaries right?

Reply:
This is a chunk of blocks, but `chunk()` returns an iterable of distinct block chunk iterables, so this is just converting a block chunk iterable into a block chunk list so we can give it to `ray.wait()` for prefetching.

> Chunks would also not be prefetching across chunk boundaries right?

Correct, as this currently stands, we'd prefetch `prefetch_blocks` blocks, process each of those blocks, then start another prefetching round. It shouldn't be difficult to restructure this such that prefetching and processing is better interleaved, although that would probably look more like a sliding window approach.